File: lib/transcribe.js
// Движок расшифровки. Единственный файл, который знает про Gemini.
//
// Интерфейс намеренно узкий: transcribe(bytes) -> {summary, transcript}.
// Если Gemini не подойдёт (гео, формат Opus, качество русского) — замена на
// Groq/whisper меняет только этот файл.

// Проверено на живом API 2026-08-01: расшифровка русской речи из OGG/Opus
// практически эталонная, 488 токенов на 15 секунд аудио.
// Выбран flash-lite ради квоты: 500 запросов в день против существенно
// меньшего лимита у 3.5-flash. Для личного бота это с запасом.
const MODEL = process.env.GEMINI_MODEL ?? "gemini-3.1-flash-lite";

// generateContent объявлен легаси, новые модели приземляются в Interactions
// API. Стартуем здесь: форма проще, примеров больше, и модель из списка
// v1beta работает.
const ENDPOINT = "https://generativelanguage.googleapis.com/v1beta/models";

const PROMPT = `Ты расшифровываешь голосовое сообщение из Telegram.

Верни JSON с двумя полями:

transcript — дословная расшифровка. Ровно то, что человек сказал.
Не приглаживай речь, не убирай повторы, не дописывай связки, которых не было.
Расставь знаки препинания и раздели на абзацы по смыслу.
Числительные записывай цифрами: «сорок две тысячи» → «42 000»,
«в три часа» → «в 15:00», если из контекста понятно время суток.
Если часть неразборчива, поставь [неразборчиво] вместо догадки.

summary — суть в двух-трёх строках. Пиши сразу по делу.
Никаких преамбул вида «в этом сообщении говорится» или «автор рассказывает».
Если в сообщении есть договорённость, дата, сумма или просьба — они должны
попасть в summary.

Отвечай по-русски, даже если в аудио есть вставки на другом языке.`;

const SCHEMA = {
  type: "OBJECT",
  properties: {
    summary: { type: "STRING" },
    transcript: { type: "STRING" },
  },
  required: ["summary", "transcript"],
};

export class TranscribeError extends Error {}
export class RateLimited extends TranscribeError {}

function apiKey() {
  const value = process.env.GEMINI_API_KEY;
  if (!value) throw new TranscribeError("GEMINI_API_KEY не задан");
  return value;
}

/**
 * @param {Buffer} audioBytes сырые байты .ogg из Telegram
 * @param {string} mimeType Gemini документирует audio/ogg как OGG Vorbis,
 *   а Telegram шлёт Opus. Если прилетит 400 на формате — это здесь.
 */
export async function transcribe(audioBytes, { mimeType = "audio/ogg", deadline } = {}) {
  const body = {
    contents: [
      {
        parts: [
          { text: PROMPT },
          {
            inline_data: {
              mime_type: mimeType,
              data: audioBytes.toString("base64"),
            },
          },
        ],
      },
    ],
    generationConfig: {
      responseMimeType: "application/json",
      responseSchema: SCHEMA,
    },
  };

  // Ретрай на разовых сбоях: сеть и 5xx. На 429 не повторяем — квота
  // исчерпана, повтор только сожжёт ещё одну попытку. На 4xx тоже не
  // повторяем: запрос не станет валиднее от повтора.
  const RETRIES = 1;
  let response;
  let lastError;

  for (let attempt = 0; attempt <= RETRIES; attempt++) {
    try {
      response = await fetch(`${ENDPOINT}/${MODEL}:generateContent`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "x-goog-api-key": apiKey(),
        },
        body: JSON.stringify(body),
        // Замерено: ~8× быстрее реального времени, то есть 10 минут аудио
        // это ~75 с. 120 с даёт полуторакратный запас. Сверху всё равно
        // висит общий дедлайн — он и есть настоящий потолок.
        signal: deadline
          ? AbortSignal.any([AbortSignal.timeout(120_000), deadline])
          : AbortSignal.timeout(120_000),
      });
    } catch (error) {
      lastError = error;
      if (deadline?.aborted) {
        throw new TranscribeError("не уложились в общий лимит времени");
      }
      if (attempt < RETRIES) {
        console.warn(`[gemini] ${error.name}, повтор`);
        await new Promise((r) => setTimeout(r, 1000));
        continue;
      }
      if (error.name === "TimeoutError" || error.name === "AbortError") {
        throw new TranscribeError("Gemini не ответил за 2 минуты, даже со второй попытки");
      }
      throw new TranscribeError(`сеть: ${error.message}`);
    }

    if (response.status >= 500 && attempt < RETRIES) {
      console.warn(`[gemini] HTTP ${response.status}, повтор`);
      await new Promise((r) => setTimeout(r, 1000));
      continue;
    }
    break;
  }

  if (response.status === 429) {
    throw new RateLimited("квота Gemini исчерпана (429)");
  }

  const raw = await response.text();

  if (!response.ok) {
    let detail;
    try {
      detail = JSON.parse(raw)?.error?.message;
    } catch {
      detail = raw.slice(0, 200);
    }
    throw new TranscribeError(`Gemini HTTP ${response.status}: ${detail}`);
  }

  const parsed = JSON.parse(raw);
  const text = parsed?.candidates?.[0]?.content?.parts?.[0]?.text;

  if (!text) {
    const reason =
      parsed?.candidates?.[0]?.finishReason ??
      parsed?.promptFeedback?.blockReason ??
      "причина неизвестна";
    throw new TranscribeError(`модель не вернула текст (${reason})`);
  }

  let payload;
  try {
    payload = JSON.parse(text);
  } catch (error) {
    throw new TranscribeError(`модель вернула не JSON: ${error.message}`);
  }

  return {
    summary: String(payload.summary ?? "").trim(),
    transcript: String(payload.transcript ?? "").trim(),
  };
}

File: lib/transcribe.rb
# Движок расшифровки. Единственная точка, которая знает про Gemini.
#
# Интерфейс намеренно узкий: Transcribe.call(bytes) -> Result.
# Если Этап 0 покажет, что Gemini не подходит (гео, формат Opus, качество
# русского) — Approach C меняет только этот файл на Groq/whisper.

require "net/http"
require "json"
require "uri"

module Transcribe
  # ВНИМАНИЕ (Open Question 5 в дизайн-доке): имя модели надо сверить в
  # AI Studio. gemini-2.5-flash устарел, 2.0-flash выключен. Держим в env,
  # чтобы менять без правки кода.
  MODEL = ENV.fetch("GEMINI_MODEL", "gemini-3.5-flash")

  # ВНИМАНИЕ (Open Question 4): generateContent объявлен легаси, новые модели
  # приземляются в Interactions API. Стартуем здесь — форма проще и примеров
  # больше — понимая, что миграция когда-нибудь будет.
  ENDPOINT = "https://generativelanguage.googleapis.com/v1beta/models"

  PROMPT = <<~PROMPT.freeze
    Ты расшифровываешь голосовое сообщение из Telegram.

    Верни JSON с двумя полями:

    transcript — дословная расшифровка. Ровно то, что человек сказал.
    Не приглаживай речь, не убирай повторы, не дописывай связки, которых не
    было. Расставь знаки препинания и раздели на абзацы по смыслу — это всё.
    Если часть неразборчива, поставь [неразборчиво] вместо догадки.

    summary — суть в двух-трёх строках. Пиши сразу по делу.
    Никаких преамбул вида «в этом сообщении говорится» или «автор рассказывает».
    Если в сообщении есть договорённость, дата, сумма или просьба — они должны
    попасть в summary.

    Отвечай по-русски, даже если в аудио есть вставки на другом языке.
  PROMPT

  SCHEMA = {
    type: "OBJECT",
    properties: {
      summary: { type: "STRING" },
      transcript: { type: "STRING" }
    },
    required: %w[summary transcript]
  }.freeze

  Result = Struct.new(:summary, :transcript, keyword_init: true)

  class Error < StandardError; end
  class RateLimited < Error; end

  module_function

  def api_key
    ENV["GEMINI_API_KEY"] or raise Error, "GEMINI_API_KEY не задан"
  end

  # audio_bytes — сырые байты .ogg из Telegram.
  # mime_type — Open Question 2: Gemini документирует audio/ogg как OGG Vorbis,
  # а Telegram шлёт Opus. Если прилетит 400 на формате, это здесь.
  def call(audio_bytes, mime_type: "audio/ogg")
    body = {
      contents: [
        {
          parts: [
            { text: PROMPT },
            {
              inline_data: {
                mime_type: mime_type,
                # pack("m0") вместо require "base64": base64 перестал быть
                # default-гемом в Ruby 3.4, а проект должен пережить апгрейд
                # рантайма без правок.
                data: [audio_bytes].pack("m0")
              }
            }
          ]
        }
      ],
      generationConfig: {
        responseMimeType: "application/json",
        responseSchema: SCHEMA
      }
    }

    parsed = post(body)
    text = extract_text(parsed)
    payload = JSON.parse(text)

    Result.new(
      summary: payload["summary"].to_s.strip,
      transcript: payload["transcript"].to_s.strip
    )
  rescue JSON::ParserError => e
    raise Error, "модель вернула не JSON: #{e.message}"
  end

  def post(body)
    uri = URI("#{ENDPOINT}/#{MODEL}:generateContent")

    request = Net::HTTP::Post.new(uri)
    request["Content-Type"] = "application/json"
    request["x-goog-api-key"] = api_key
    request.body = JSON.generate(body)

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true
    http.open_timeout = 10
    # Держим ниже таймаута вебхука Telegram (~60 с): лучше своя внятная
    # ошибка, чем ретрай Telegram и дубль ответа пользователю.
    http.read_timeout = 45

    response = http.request(request)

    case response
    when Net::HTTPSuccess
      JSON.parse(response.body)
    when Net::HTTPTooManyRequests
      raise RateLimited, "квота Gemini исчерпана (429)"
    else
      detail = begin
        JSON.parse(response.body).dig("error", "message")
      rescue StandardError
        nil
      end
      raise Error, "Gemini HTTP #{response.code}: #{detail || response.body[0, 200]}"
    end
  rescue Net::OpenTimeout, Net::ReadTimeout
    raise Error, "Gemini не ответил вовремя"
  end

  def extract_text(parsed)
    text = parsed.dig("candidates", 0, "content", "parts", 0, "text")
    return text if text && !text.empty?

    reason = parsed.dig("candidates", 0, "finishReason") ||
             parsed.dig("promptFeedback", "blockReason")
    raise Error, "модель не вернула текст (#{reason || "причина неизвестна"})"
  end
end
